javascript/extensible-search-typeahead.js

```diff
@@ -0,0 +1,229 @@
+;(function($) {
+	window.ESPPageStore = new PageStore;
+	$(window).load(function() {
+
+		$.fn.extend({
+			nearest: function (pattern) {
+
+				var found = null;
+				var domobj = null;
+
+				do {
+					domobj = this.parent();
+					found = domobj.find(pattern);
+				} while (found === null && domobj !== null)
+
+				return found;
+			}
+		});
+
+
+		var overlays = $('.esp-overlay');
+
+		if(overlays.length) {
+
+			bindSearchCapture();
+
+			overlays.each(function (i , overlay) {
+				overlay =  $(overlay);
+				var search = overlay.nearest('input.extensible-search.typeahead');
+				if(search.length) {
+
+					search.after(overlay);
+
+					if (ESPPageStore.getType('SearchTerms').length && overlay.children('.recent-searches').length) {
+
+						var resentsearch = overlay.find('.recentsearch').first().clone();
+						overlay.find('.recentsearch').remove();
+
+						overlay.find('.search-suggestions-list > li.list-item > a')
+							.each(function(){
+								$(this).attr('href',
+									'//' + window.location.host
+									+ $(this).parents('form').attr('action')
+									+ '?Search=' + $(this).text()
+								);
+							});
+
+						$.each(ESPPageStore.getType('SearchTerms'), function (i, term) {
+							overlay.find('.recent-searches-list').prepend(
+								resentsearch.clone()
+									.children('a')
+									.prop('href',term.link)
+									.text(term.title)
+									.parent()
+									.show()
+							);
+						});
+
+						$('.recent-searches').show();
+
+					}
+
+					search.keyup(function() {
+						var currSearch = $(this);
+						var data = {'page': currSearch.data('extensible-search-page')};
+						if(currSearch.val().length > 2) {
+							var list = overlay.find('.search-typeahead-list');
+							searchAjax(currSearch, data, list, 'extensible-search-api/getTypeahead');
+							var list = overlay.find('.search-suggestions-list');
+							searchAjax(currSearch, data, list, 'extensible-search-api/getSuggestions');
+						}
+					})
+					.on({
+						focus: function () {
+							showOverlay(search, overlay);
+						}
+					});
+				}
+			});
+
+		}
+
+	});
+
+	function showOverlay(search, overlay) {
+		overlay.show();
+		search.is();
+		$(document).on("focusin click", function () {
+			// Are we clicking or focused anywhere in the overlay or search
+			if (overlay.has($(event.target)).length && $(event.target).is('a')) {
+				//If the focus is a link don't hide the overlay so we can tab through them
+				return;
+			} else if ($(event.target).is(overlay)
+					|| $(event.target).is(search)
+					|| overlay.has($(event.target)).length
+			) {
+				//else the user has missed the links but hit the overlay return focus to the search input
+				//Prevent event propagation to prevent an endless loop
+				search.focus(function( event ) {
+					event.stopPropagation();
+				});
+				search.focus();
+				return;
+			}
+			overlay.hide();
+		});
+	}
+
+	function searchAjax(currSearch, data, list, endpoint) {
```

Comment:
Assuming it's possible to unify the two above ajax endpoints, you should only be calling the searchAjax function _if_ there's not a pending request running (to avoid spamming the server with requests that are not ever going to have an end effect). So, store the promise returned by $.get in a variable, that gets set to null on request completion. If that variable is _not_ null, then you shouldn't be triggering the $.get call, and instead just return the existing promise.